File: Data Structures & Algorithms/depth-of-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        queue = deque()
        d=0
        queue.append(root)
        while queue:
            n=len(queue)
            for i in range(n):
                node=queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            d+=1
        # using bfs TC O(n) and SC(n)
        return d

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Pair struct {
    Node  *TreeNode
    V     int
}

type Queue struct{
    element []Pair
}

func (q *Queue) enqueue(p Pair){
    q.element = append(q.element,p)
}

func (q *Queue) dequeue() Pair{
    if len(q.element) == 0{
        return Pair {}
    }

    res :=q.element[0]
    q.element=q.element[1:]
    return res
}

func (q *Queue) queSize() int{
    return len(q.element)
}

func goodNodes(root *TreeNode) int {
    if root == nil{
        return 0
    }
    var res int
    q := Queue{}

    q.enqueue(Pair{Node:root,V:math.MinInt})

    for q.queSize()>0{

        front:=q.dequeue()
        curr, maxVal:=front.Node, front.V

        if curr.Val >= maxVal{
            res++
            maxVal = curr.Val
        }
        if curr.Left != nil{
            q.enqueue(Pair{Node:curr.Left,V:maxVal})
        }
        if curr.Right !=nil{
            q.enqueue(Pair{Node:curr.Right,V:maxVal})
        }
        
    }
    return res
}
